### PR TITLE
Enable optimization for conv2d in HiFi4 (#772)

### PR DESCRIPTION
**Pull request by tingyan19**
Apply patches to enable optimization for conv2d in HiFi4

BUG=#772

TESTED=

docker run --rm -v "$(pwd)":/opt/tflite-micro ghcr.io/tflm-bot/xtensa:0.1 /opt/tflite-micro/tensorflow/lite/micro/tools/ci_build/test_xtensa_hifi3z.sh

### CODE REVIEW

File: tensorflow/lite/micro/kernels/xtensa/conv_hifi.cc [hunk 1]
```diff
@@ -54,23 +54,142 @@ TfLiteStatus ConvPrepareHifi(TfLiteContext* context, TfLiteNode* node) {
...
+                            const TfLiteEvalTensor* filter,
+                            const TfLiteEvalTensor* bias,
+                            TfLiteEvalTensor* output) {
+  const RuntimeShape& input_shape = tflite::micro::GetTensorShape(input);
+  const RuntimeShape& filter_shape = tflite::micro::GetTensorShape(filter);
+  /* Dilation is currently not supported on HiFi 4 NN Library */
+  if ((params.dilation_width_factor == 1) &&
+      (params.dilation_height_factor == 1) &&
+      input_shape.Dims(1) >= filter_shape.Dims(1) &&
+      input_shape.Dims(2) >= filter_shape.Dims(2)) {
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/xtensa/conv_hifi.cc:93]: this if-else block is quite confusing to read.

The else block has a return statement and then there is a call to the reference kernel after the else block. Refactoring to something like the following might help:

```cc
if (input_shape.Dims(1) < filter_shape.Dims(1)  || input_shape.Dims(2) >= filter_shape.Dims(2)) {
// first set of calls to nnlib
return
}

// Dilation not supported in nnlib
if ((params.dilation_width_factor == 1) &&
      (params.dilation_height_factor == 1) &&
      input_shape.Dims(1) >= filter_shape.Dims(1) &&
      input_shape.Dims(2) >= filter_shape.Dims(2)) {
 // calls to nnlib
  return;
}

// Use the reference kernels
}
```

I would also like know if the nnlib call when filter is large than the input is necessary, and whether that supports dilation or not.

File: tensorflow/lite/micro/tools/make/Makefile [hunk 2]
```diff
@@ -112,6 +112,7 @@ endif
 
 CORE_OPTIMIZATION_LEVEL := -Os
 KERNEL_OPTIMIZATION_LEVEL := -O2
+THIRD_PARTY_C_OPTIMIZATION_LEVEL := -Os
```
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/Makefile:115]: Let's keep this change for a separate PR. I think we will want to be careful here since we might want some 3P libraries with -Os (e.g. flatbuffers) and others with -O2 (CMSIS, nnlib). Perhaps make a bug and we can discuss further?

Related bug: http://b/161489252 and https://github.com/tensorflow/tflite-micro/issues/638

File: tensorflow/lite/micro/tools/make/targets/xtensa_makefile.inc [hunk 3]
```diff
@@ -34,6 +34,8 @@ endif
 #   TARGET_ARCH=hifi4 --> -DHIFI4
 TARGET_ARCH_DEFINES := -D$(shell echo $(TARGET_ARCH) | tr [a-z] [A-Z])
 
+THIRD_PARTY_C_OPTIMIZATION_LEVEL := -O3 -LNO:simd
```
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/xtensa_makefile.inc:37]: let's keep this change for a separate PR

**Review by advaitjain — CHANGES_REQUESTED** [review]

File: tensorflow/lite/micro/tools/make/targets/xtensa_makefile.inc (hunk 3 above)
Comment (tingyan19) [reply tensorflow/lite/micro/tools/make/targets/xtensa_makefile.inc]: Rolled back in commit https://github.com/tensorflow/tflite-micro/pull/774/commits/90c7ba043d30dc2879e5bb82e8e62c34f0be4418

and created bug: 

http://b/210160638

File: tensorflow/lite/micro/tools/make/Makefile (hunk 2 above)
Comment (tingyan19) [reply tensorflow/lite/micro/tools/make/Makefile]: Rolled back in commit https://github.com/tensorflow/tflite-micro/commit/90c7ba043d30dc2879e5bb82e8e62c34f0be4418

and created bug:

http://b/210160638

File: tensorflow/lite/micro/kernels/xtensa/conv_hifi.cc (hunk 1 above)
Comment (tingyan19) [reply tensorflow/lite/micro/kernels/xtensa/conv_hifi.cc:93]: Created an cleanup bug http://b/210163787 to refactor it soon.

**Review by advaitjain — DISMISSED** [review]

**Review by advaitjain — APPROVED** [review]
